src/llmq/quorums_signing_shares.cpp

```diff
@@ -0,0 +1,1182 @@
+            return;
+        }
+        
+        sigSharesToAnnounce.emplace(sigShare.GetKey());
+        firstSeenForSessions.emplace(sigShare.GetSignHash(), GetTimeMillis());
+
+        if (!quorumNodes.empty()) {
+            // don't announce and wait for other nodes to request this share and directly send it to them
+            // there is no way the other nodes know about this share as this is the one created on this node
+            // this will also indicate interest to the other nodes in sig shares for this quorum
+            for (auto& p : nodeStates) {
+                if (!quorumNodes.count(p.first) && !p.second.interestedIn.count(std::make_pair((Consensus::LLMQType)sigShare.llmqType, sigShare.quorumHash))) {
+                    continue;
+                }
+                p.second.MarkRequested((Consensus::LLMQType)sigShare.llmqType, sigShare.GetSignHash(), sigShare.quorumMember);
+                p.second.MarkKnows((Consensus::LLMQType)sigShare.llmqType, sigShare.GetSignHash(), sigShare.quorumMember);
+            }
+        }
+
+        size_t sigShareCount = CountBySignHash(sigShares, sigShare.GetSignHash());
+        if (sigShareCount >= quorum->params.threshold) {
+            canTryRecovery = true;
+        }
+    }
+
+    if (canTryRecovery) {
+        TryRecoverSig(quorum, sigShare.id, sigShare.msgHash, connman);
+    }
+}
+
+void CSigSharesManager::TryRecoverSig(const CQuorumCPtr& quorum, const uint256& id, const uint256& msgHash, CConnman& connman)
+{
+    if (quorumSigningManager->HasRecoveredSigForId(quorum->params.type, id)) {
+        return;
+    }
+
+    std::vector<CBLSSignature> sigSharesForRecovery;
+    std::vector<CBLSId> idsForRecovery;
+    {
+        LOCK(cs);
+
+        auto k = std::make_pair(quorum->params.type, id);
+
+        auto signHash = CLLMQUtils::BuildSignHash(quorum->params.type, quorum->quorumHash, id, msgHash);
+        auto itPair = FindBySignHash(sigShares, signHash);
+
+        sigSharesForRecovery.reserve((size_t) quorum->params.threshold);
+        idsForRecovery.reserve((size_t) quorum->params.threshold);
+        for (auto it = itPair.first; it != itPair.second && sigSharesForRecovery.size() < quorum->params.threshold; ++it) {
+            auto& sigShare = it->second;
+            sigSharesForRecovery.emplace_back(sigShare.sigShare);
+            idsForRecovery.emplace_back(CBLSId::FromHash(quorum->members[sigShare.quorumMember]->proTxHash));
+        }
+
+        // check if we can recover the final signature
+        if (sigSharesForRecovery.size() < quorum->params.threshold) {
+            return;
+        }
+    }
+
+    // now recover it
+    cxxtimer::Timer t(true);
+    CBLSSignature recoveredSig;
+    if (!recoveredSig.Recover(sigSharesForRecovery, idsForRecovery)) {
+        LogPrintf("CSigSharesManager::%s -- failed to recover signature. id=%s, msgHash=%s, time=%d\n", __func__,
+                  id.ToString(), msgHash.ToString(), t.count());
+    } else {
```

Comment:
nit:
```suggestion
        return;
    }
```
and make corresponding changes below